src/Accounts/Authentication/Utilities/FrequencyService.cs

```diff
@@ -0,0 +1,194 @@
+        public class FrequencyInfo
+        {
+            public TimeSpan Frequency { get; set; }
+            public DateTime LastCheckTime { get; set; }
+            public FrequencyInfo(TimeSpan frequency, DateTime lastCheckTime)
+            {
+                Frequency = frequency;
+                LastCheckTime = lastCheckTime;
+            }
+        }
+
+        public FrequencyService(IDataStore dataStore)
+        {
+            try
+            {
+                _dataStore = dataStore;
+                _clock = new Clock();
+                if (dataStore.FileExists(_filePath))
+                {
+                    string json = dataStore.ReadFileAsText(_filePath);
+                    _frequencies = JsonConvert.DeserializeObject<Dictionary<string, FrequencyInfo>>(json);
+                }
+                else
+                {
+                    _frequencies = new Dictionary<string, FrequencyInfo>();
+                }
+            }
+            catch (Exception)
+            {
+                _dataStore = new MemoryDataStore();
+
+                _frequencies = new Dictionary<string, FrequencyInfo>();
+
+            }
+
+        }
+
+        public FrequencyService(IDataStore dataStore, IClock clock)
+        {
+            try
+            {
+                _dataStore = dataStore;
+                _clock = clock;
+                if (dataStore.FileExists(_filePath))
+                {
+                    string json = dataStore.ReadFileAsText(_filePath);
+                    _frequencies = JsonConvert.DeserializeObject<Dictionary<string, FrequencyInfo>>(json);
+                }
+                else
+                {
+                    _frequencies = new Dictionary<string, FrequencyInfo>();
+                }
+            }
+            catch (Exception ex)
+            {
+                throw new Exception("Error loading frequency file", ex);
+            }
+        }
+        public void TryRun(string featureName, Func<bool> businessCheck, Action business)
+        {
+            if (_perPSSessionRegistry != null && _perPSSessionRegistry.ContainsKey(featureName))
+            {
+                if (_perPSSessionRegistry[featureName] == false && businessCheck())
+                {
+                    _perPSSessionRegistry[featureName] = true;
+                    business();
+                }
+            }
+            else if (_frequencies.ContainsKey(featureName))
+            {
+                if (_clock.IsDue(_frequencies[featureName].LastCheckTime, _frequencies[featureName].Frequency) && businessCheck())
+                {
+                    _frequencies[featureName].LastCheckTime = DateTime.Now;
+                    business();
+                    Save();
+                }
+            }
+            else
+            {
+                throw new ArgumentException($"Feature name '{featureName}' not found in FrequencyService");
```

Comment:
Change Exception to AzPs*Exception, see https://github.com/Azure/azure-powershell-common/tree/main/src/Common/Exceptions